src/main/java/net/neoforged/neoforge/codec/NBTSerializableCodec.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.codec;
+
+import static net.neoforged.neoforge.codec.NeoforgeNbtCodecs.TAG_CODEC;
+
+import com.mojang.datafixers.util.Pair;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.DynamicOps;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.util.Arrays;
+import java.util.function.Supplier;
+import net.minecraft.core.HolderLookup;
+import net.minecraft.nbt.Tag;
+import net.minecraft.resources.RegistryOps;
+import net.neoforged.neoforge.common.util.INBTSerializable;
+
+/**
+ * A codec for instances of {@link INBTSerializable}.
+ *
+ * @param defaultSupplier Used to construct empty instances.
+ * @param serializerClass The class implementing {@link INBTSerializable}.
+ * @param nbtTypeClass    The type of NBT tag the serializer serializes into.
+ * @param serializer      Method reference ({@link INBTSerializable#serializeNBT(HolderLookup.Provider)})
+ * @param deserializer    Method reference ({@link INBTSerializable#deserializeNBT(HolderLookup.Provider, Tag)})
+ * @param <S>             Serializer type param.
+ * @param <NBT>           NBT type param.
+ */
+public record NBTSerializableCodec<S extends INBTSerializable<NBT>, NBT extends Tag>(Supplier<S> defaultSupplier,
+        Class<S> serializerClass,
+        Class<NBT> nbtTypeClass,
+        Method serializer,
+        Method deserializer) implements Codec<S> {
+    public static <S extends INBTSerializable<NBT>, NBT extends Tag> Codec<S> forSerializable(Class<S> serializerClass, Supplier<S> defaultSupplier) {
+        try {
+            final var nbtTypeClass = Arrays.stream(serializerClass.getGenericInterfaces())
```

Comment:
_What_ the actual heck is going on with this codec. Why are we doing all this reflectively? This whole thing seems... just, _really_ fricking iffy and a sign something is wrong with the API. The parameterized type checks are fragile. The reflective stuff is asking for pain down the road. I'm entirely uncertain why we're doing this by passing the _class_ in at all -- should be enough to pass in `SerializedClass::deserializeNBT` as a lambda and the supplier and have the rest be handled in a typesafe fashion by just normal old DFU techiniques, making use of the generics for compile-time type safety instead of relying on runtime checks?